tensorflow/core/kernels/dml_topk_op.cc

```diff
@@ -0,0 +1,145 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+Portions Copyright (c) Microsoft Corporation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/bounds_check.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class DmlTopKInitHelper : public InitializationHelper {
+ public:
+  struct Attributes {
+    explicit Attributes(OpKernelConstruction* ctx) {
+      OP_REQUIRES_OK(ctx, ctx->GetAttr("sorted", &sorted_));
+      if (ctx->num_inputs() < 2) {  // k is an attr (TopK).
+        OP_REQUIRES_OK(ctx, ctx->GetAttr("k", &k_));
+      } else {  // k is an input (TopKV2), so we won't know it until Compute.
+        k_ = -1;
+      }
+    }
+    int k_;
+    bool sorted_;
+  };
+
+  DmlTopKInitHelper(OpKernelContext* ctx,
+                    std::shared_ptr<const Attributes> attr)
+      : attr_(std::move(attr)) {
+    k = attr_->k_;
+    if (ctx->num_inputs() >= 2) {
+      const auto& k_in = ctx->input(1);
+      OP_REQUIRES(ctx, TensorShapeUtils::IsScalar(k_in.shape()),
+                  errors::InvalidArgument("k must be scalar, got shape ",
+                                          k_in.shape().DebugString()));
+      k = k_in.scalar<int32>()();
+    }
+    OP_REQUIRES(ctx, k >= 0, errors::InvalidArgument("Need k >= 0, got ", k));
+    const auto& input_in = ctx->input(0);
+    OP_REQUIRES(ctx, input_in.dims() >= 1,
+                errors::InvalidArgument("input must be >= 1-D, got shape ",
+                                        input_in.shape().DebugString()));
+    OP_REQUIRES(ctx, input_in.dim_size(input_in.dims() - 1) >= k,
+                errors::InvalidArgument(
+                    "input must have at least k columns. Had ",
+                    input_in.dim_size(input_in.dims() - 1), ", needed ", k));
+  }
+  bool GetSorted() const { return attr_->sorted_; }
```

Comment:
`GetSorted` isn't actually used below.

Reply:
That's true, the order of output of topk(..., sorted=False) is not defined in tensorflow. The test will just verify the output set. I'll delete the methods to return sorted_.

---
In reply to: [491183198](https://github.com/microsoft/tensorflow-directml/pull/44#discussion_r491183198) [](ancestors = 491183198)